browser/src/layer/tile/CanvasTileWorker.js

```diff
@@ -0,0 +1,116 @@
+/* -*- js-indent-level: 8; fill-column: 100 -*- */
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/* eslint-disable no-inner-declarations */
+/* eslint no-unused-vars: ["warn", { "argsIgnorePattern": "^_" }] */
+/* global importScripts Uint8Array */
+
+if ('undefined' === typeof window) {
+	self.L = {};
+
+	importScripts('../../../node_modules/fzstd/umd/index.js');
+	importScripts('CanvasTileUtils.js');
+	addEventListener('message', onMessage);
+
+	console.info('CanvasTileWorker initialised');
+
+	var keyframeBufferSize = 0;
+	var keyframeBuffer;
+
+	function onMessage(e) {
+		switch (e.data.message) {
+			case 'decompress':
+				var buffer = e.data.buffer;
+				var tile = e.data.tile;
+
+				// Allocate key-frame decompression buffer
+				// FIXME: This is an over-allocation, we should store the uncompressed size
+				//        on the server-side and communicate that with the client.
+				var tileByteSize = tile.tileSize * tile.tileSize * 4;
+				if (tileByteSize > keyframeBufferSize) {
+					keyframeBufferSize = tileByteSize;
+					keyframeBuffer = new Uint8Array(tileByteSize);
+				}
+
+				var fzstdBuffer = tile.isKeyframe ? keyframeBuffer : buffer;
+				var fzstdOffset = 0;
+
+				var stream = new self.fzstd.Decompress((chunk, _isLast) => {
+					fzstdBuffer.set(chunk, fzstdOffset);
+					fzstdOffset += chunk.length;
+				});
+				stream.push(tile.rawDelta);
```

Comment:
I believe this is a good performance optimisation that could apply to the source tree without workers - pre-allocating the decompression buffer rather than constantly creating and deleting them. I was seeing what looked like memory fragmentation causing GC pauses in profiles I did earlier in the year. The same goes for pre-allocating and recycling tile buffers.